Defer Script Host disposal during cold start

balag0 opened the pull request:
Defer Script Host disposal during cold start

One of the steps during specialization is to restart host. This restart involves starting a new host and disposing the old host.

From profiling, disposal of the old host is the most expensive part in the restart operation. On a single core machine, while the restart is in progress, function invocation will be delayed during cold start. On most cold starts the host dispose seems to be getting scheduled only after cold start has completed so this doesn't impact cold start at lower percentiles. But if the ordering of execution happened to be reversed, the host dispose will increase cold start.

Fix is to defer the Host disposal by 5 seconds. The change reduces P90 cold start by 60%.

<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves https://github.com/Azure/azure-functions-host/issues/7305

### Pull request checklist

* [X ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ X] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ X] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

fabiocav commented on src/WebJobs.Script.WebHost/Helpers/ScheduledDisposer.cs:
```diff
@@ -0,0 +1,139 @@
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"{nameof(ScheduleTimer)}");
```
Let's not swallow this exception here. This might well be something we can't recover from.

balag0 replied:
Good point. Will fix this.

balag0 replied:
removed this code

fabiocav commented on src/WebJobs.Script.WebHost/Helpers/SemaphoreHelper.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public class SemaphoreHelper
+    {
+        public virtual bool Wait(SemaphoreSlim semaphore, int millisecondsTimeout)
```
Can you clarify why this class is needed? Is this purely to avoid registering a semaphore?

balag0 replied:
Its to make unit testing easier.

balag0 replied:
not used anymore.

fabiocav reviewed (commented):
This change could have unintended side effects and it introduces enough of a behavior change that it could impact customers. Have you considered those? Was there a consideration on having this pushed to 4.0 instead?

balag0 commented:
> This change could have unintended side effects and it introduces enough of a behavior change that it could impact customers. Have you considered those? Was there a consideration on having this pushed to 4.0 instead?

Thanks for taking a look @fabiocav 
Yes, that is what I am trying to make sure with this PR :). Could you explain a bit more on how this impacts customers. I am not very familiar with this code and looking at the way things are handled currently (fire and forget) it doesn't seem like we are enforcing any ordering with this task.  Also the exact ordering being enforced by this change is already happening in prod every now and then. So not clear on what the behavior change is :)

balag0 commented:
Updated the PR to use ExecuteAfterColdStartDelay thanks to @safihamid's suggestion

balag0 commented:
> > This change could have unintended side effects and it introduces enough of a behavior change that it could impact customers. Have you considered those? Was there a consideration on having this pushed to 4.0 instead?
> 
> Thanks for taking a look @fabiocav
> Yes, that is what I am trying to make sure with this PR :). Could you explain a bit more on how this impacts customers. I am not very familiar with this code and looking at the way things are handled currently (fire and forget) it doesn't seem like we are enforcing any ordering with this task. Also the exact ordering being enforced by this change is already happening in prod every now and then. So not clear on what the behavior change is :)

@fabiocav @brettsam I have updated the PR to make the changes simpler. Please do take a look and let me know if (and how) this will have any customer impact :)

balag0 commented on src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs:
```diff
@@ -134,6 +135,9 @@ public static class WebScriptHostBuilderExtension
                     services.TryAddEnumerable(ServiceDescriptor.Singleton<IHostedService, IFileMonitoringService>(p => p.GetService<IFileMonitoringService>()));
 
+                    // Record Standby state when the Host was created
+                    services.AddSingleton(new ScriptHostStandbyStateProvider(standbyOptions.InStandbyMode));
```
Also considered injecting ScriptHostStandbyStateProvider to ScriptHost and exposing this with a getter on ScriptHost. But this seemed like a smaller changelist since this is not actually needed in ScriptHost and some of the tests were complicated to write since it is not easy to mock  ScriptHost.

brettsam replied:
Rather than introduce a new service, could you add an `IsStandbyConfiguration` property to `ScriptJobHostOptions`? In the `ScriptHostOptionsSetup` (which should be renamed to match the options) -- it could copy the value directly from the `ScriptApplicationHostOptions`, which already has an `IsStandbyConfiguration` property.

balag0 replied:
done.

brettsam commented on src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:
```diff
@@ -683,7 +691,31 @@ private async Task Orphan(IHost instance, CancellationToken cancellationToken =
+                        {
+                            try
+                            {
+                                _logger.LogDebug("Starting ScripHost dispose");
```
Should these calls be using `GetHostLogger()`?

brettsam replied:
Also, can we make these two messages specifically mention that they're disposing the standby host? That'll be easier to diagnose things in kusto.

balag0 replied:
fixed.

brettsam reviewed (commented):
One refactoring comment that may clean up the PR a bit.

pragnagopa commented on src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:
```diff
@@ -684,7 +686,31 @@ private async Task Orphan(IHost instance, CancellationToken cancellationToken =
+                            }
+                            catch (Exception e)
+                            {
+                                _logger.LogError(e, "Failed to dispose Standby ScriptHost instance");
```
is try...catch needed? Ok to keep running even this disposal fails?

brettsam replied:
i think it's okay to catch and log -- but we should re-throw the exception to match the previous behavior (of not catching).

balag0 replied:
Thanks. Yes didnt mean to change that behavior.

balag0 replied:
yes. didn't mean to change that behavior. Fixed.

pragnagopa reviewed (commented):
Changes LGTM! Scoped to cold start request only! 

@fabiocav / @brettsam  - any other concerns?

brettsam commented on src/WebJobs.Script/Config/ScriptJobHostOptionsSetup.cs:
```diff
@@ -8,7 +8,7 @@
 namespace Microsoft.Azure.WebJobs.Script.Configuration
 {
-    internal class ScriptHostOptionsSetup : IConfigureOptions<ScriptJobHostOptions>
+    internal class ScriptJobHostOptionsSetup : IConfigureOptions<ScriptJobHostOptions>
```
Thank you! :-)

balag0 replied:
:)

balag0 replied:
:)

brettsam reviewed (commented):
Just 1 minor comment about that try/catch.

balag0 committed:
Defer Script Host disposal during cold start

    One of the steps during specialization is to restart host. This restart involves starting a new host and disposing the old host.

    From profiling, disposal of the old host is the most expensive part in the restart operation. On a single core machine, while the restart is in progress, function invocation will be delayed during cold start. On most cold starts the host dispose seems to be getting scheduled only after cold start has completed so this doesn't impact cold start at lower percentiles. But if the ordering of execution happened to be reversed, the host dispose will increase cold start.

    Fix is to defer the Host disposal by 5 seconds. The change reduces P90 cold start by 60%.
